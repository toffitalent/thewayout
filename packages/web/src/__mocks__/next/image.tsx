import * as React from 'react';

// eslint-disable-next-line react/jsx-no-useless-fragment
const mock = (props: { children: React.ReactElement }): React.ReactElement => <>{props.children}</>;

export default mock;
