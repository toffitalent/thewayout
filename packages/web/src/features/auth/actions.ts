import { CreateUserRequest } from '@two/shared';
import { API } from '@app/api';
import { createApiThunk } from '@app/utils';

export const fetchUser = createApiThunk('auth/fetchUser', API.user.retrieve);
export const forgotPassword = createApiThunk('auth/forgotPassword', API.user.forgotPassword);

export const login = createApiThunk(
  'auth/login',
  async (credentials: Parameters<typeof API.login>[0], { dispatch }) => {
    await API.login(credentials);
    await dispatch(fetchUser());
  },
);

export const logout = createApiThunk('auth/logout', API.logout);

export const resetPassword = createApiThunk(
  'auth/resetPassword',
  async (
    { username, password, token }: { username: string; password: string; token: string },
    { dispatch },
  ) => {
    await API.user.resetPassword(password, token);
    return dispatch(login({ username, password }));
  },
);

export const signUp = createApiThunk(
  'auth/register',
  async (payload: CreateUserRequest, { dispatch }) => {
    await API.user.create(payload);
    return dispatch(login({ username: payload.email, password: payload.password }));
  },
);

export const updateUser = createApiThunk('auth/updateUser', API.user.update);
