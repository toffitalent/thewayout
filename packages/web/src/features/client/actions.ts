import { ListRspClientRequest } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createClient = createApiThunk('client/create', API.clientProfile.create);

export const updateClient = createApiThunk('client/update', API.clientProfile.update);

export const listAllJobs = createApiThunk(
  'client/jobs',
  ({ forceReload }: { forceReload?: boolean }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.jobsList;
    const lastId = [...ids].pop() as string;
    return API.job.list({
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const retrieveJob = createApiThunk(
  'client/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const applyJob = createApiThunk(
  'client/jobApply',
  ({ jobId, questions }: { jobId: string; questions?: { [key: string]: string }[] }) =>
    API.job.apply(jobId, {
      ...(questions ? { questions } : {}),
    }),
);

export const listAppliedJobs = createApiThunk(
  'client/applications',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).clientJobs.appliedJobs;
    const lastId = [...ids].pop() as string;
    return API.clientProfile.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const listSuggestedJobs = createApiThunk('client/suggestedJobs', (clientId: string) =>
  API.clientProfile.listSuggestedJobs(clientId),
);

export const listRsp = createApiThunk(
  'rsp/list',
  ({ userId, ...rest }: ListRspClientRequest & { userId: string }) =>
    API.clientProfile.listRsp({ userId, ...rest }),
);

export const retrieveCaseManager = createApiThunk(
  'client/caseManager',
  API.clientProfile.getCaseManager,
);
