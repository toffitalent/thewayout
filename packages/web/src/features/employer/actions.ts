import { Job, JobApplicationStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createEmployer = createApiThunk('employer/create', API.employer.create);

export const updateEmployer = createApiThunk('employer/update', API.employer.update);

export const createJob = createApiThunk('employer/job/create', API.job.create);

export const listJobs = createApiThunk(
  'employer/jobs',
  ({ forceReload, userId }: { forceReload?: boolean; userId: string }, { getState }) => {
    const { ids } = (getState() as State).employerJobs.jobs;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      ...(!forceReload && { before: lastId }),
    });
  },
);

export const updateListJobs = createApiThunk(
  'employer/jobs/listUpdated',
  ({ userId }: { userId: string }, { getState }) => {
    const { listUpdatedAt, ids } = (getState() as State).employerJobs.jobs;
    const firstId = ids[0] as string;
    const lastId = [...ids].pop() as string;
    return API.employer.listJobs(userId, {
      since: listUpdatedAt,
      after: firstId,
      before: lastId,
    });
  },
);

export const retrieveJob = createApiThunk(
  'employer/job',
  ({ jobId, since }: { jobId: string; since?: Date }) =>
    API.job.retrieve(jobId, {
      ...(since ? { since } : {}),
    }),
);

export const listApplications = createApiThunk(
  'employer/listApplications',
  ({ employerId, jobId }: { employerId: string; jobId: string }, { getState }) => {
    const { entities } = (getState() as State).employerJobs.jobsApplications;
    const page = entities[jobId]?.page;
    const isListEmpty = !entities[jobId]?.applications.length;

    return API.employer.listApplications(employerId, jobId, {
      page: isListEmpty ? 0 : page! + 1,
    });
  },
);

export const getClient = createApiThunk(
  'employer/getClient',
  ({ employerId, jobId, clientId }: { employerId: string; jobId: string; clientId: string }) =>
    API.employer.getClient(employerId, jobId, clientId, {}),
);

export const updateJob = createApiThunk(
  'employer/updateJob',
  ({ jobId, patch }: { jobId: string; patch: Partial<Job> }) => API.job.update(jobId, patch),
);

export const interviewClient = createApiThunk(
  'employer/interviewClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.interview,
    ),
);

export const hireClient = createApiThunk(
  'employer/hireClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.hired,
    ),
);

export const rejectClient = createApiThunk(
  'employer/rejectClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.rejected,
    ),
);

export const notAFitClient = createApiThunk(
  'employer/notAFitClient',
  ({
    employerId,
    jobId,
    applicationId,
  }: {
    employerId: string;
    jobId: string;
    applicationId: string;
  }) =>
    API.employer.changeApplicationStatus(
      employerId,
      jobId,
      applicationId,
      JobApplicationStatus.notAFit,
    ),
);
