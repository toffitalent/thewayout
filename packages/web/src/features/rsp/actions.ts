import { ClientType, RspClientStatus } from '@two/shared';
import { API } from '@app/api';
import type { State } from '@app/store';
import { createApiThunk } from '@app/utils';

export const createRsp = createApiThunk('rsp/create', API.rsp.create);

export const updateRsp = createApiThunk('rsp/update', API.rsp.update);

export const createMember = createApiThunk('rsp/createMember', API.rsp.createMember);

export const inviteCaseManager = createApiThunk('rsp/inviteCaseManager', API.rsp.inviteCaseManager);

export const removeInvitation = createApiThunk('rsp/removeInvitation', API.rsp.removeInvitation);

export const removeCaseManager = createApiThunk('rsp/removeCaseManager', API.rsp.removeCaseManager);

export const listCaseManagers = createApiThunk(
  'rsp/caseManagers',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.caseManagers;
    const lastId = [...ids].pop() as string;
    return API.rsp.listCaseManagers(rspId, {
      before: lastId,
    });
  },
);

export const listInvitations = createApiThunk(
  'rsp/invitations',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.invitations;
    const lastId = [...ids].pop() as string;
    return API.rsp.listInvitations(rspId, {
      before: lastId,
    });
  },
);

export const listCaseManagersNames = createApiThunk(
  'rsp/listCaseManagersNames',
  API.rsp.listCaseManagersNames,
);

export const retrieveRspAccount = createApiThunk('rsp/retrieveRspAccount', API.rsp.retrieveAccount);

export const updateCaseManager = createApiThunk('rsp/updateCaseManager', API.rsp.updateCaseManager);

export const listClients = createApiThunk(
  'rsp/clients',
  ({ rspId }: { rspId: string }, { getState }) => {
    const { ids } = (getState() as State).rsp.clients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      before: lastId,
    });
  },
);

export const listStatisticsClients = createApiThunk(
  'rsp/statisticsClients',
  (
    {
      rspId,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      forceReload,
      ...clientFilters
    }: Partial<ClientType> & { rspId: string; forceReload?: boolean },
    { getState },
  ) => {
    const { ids } = (getState() as State).rsp.statisticsClients;
    const lastId = [...ids].pop() as string;
    return API.rsp.listClients(rspId, {
      ...clientFilters,
      before: lastId,
    });
  },
);

export const retrieveClient = createApiThunk('rsp/retrieveClient', API.rsp.retrieveClient);

export const declineClient = createApiThunk(
  'rsp/declineClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.closed),
);

export const acceptClient = createApiThunk(
  'rsp/acceptClient',
  ({ rspId, clientId }: { rspId: string; clientId: string }) =>
    API.rsp.changeClientStatus(rspId, clientId, RspClientStatus.active),
);

export const addClientNotes = createApiThunk('rsp/addClientNotes', API.rsp.addClientNotes);

export const assignCaseManager = createApiThunk(
  'rsp/assignCaseManager',
  ({
    rspId,
    clientId,
    caseManagerId,
  }: {
    rspId: string;
    clientId: string;
    caseManagerId: string;
  }) => API.rsp.assignCaseManager(rspId, clientId, caseManagerId),
);

export const getStatistics = createApiThunk('rsp/statistics', API.rsp.getStatistics);
