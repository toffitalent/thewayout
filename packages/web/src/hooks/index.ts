export * from './useAppAuth';
export * from './useAppDispatch';
export * from './useAppSelector';
export * from './useDashboardRedirection';
export * from './useErrorIndicator';
export * from './useFormattedPhoneNumber';
export * from './useIndicators';
export * from './useMediaQuery';
export * from './usePrevious';
export * from './useTimeout';
