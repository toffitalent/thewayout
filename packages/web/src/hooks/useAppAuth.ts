import { useAuth } from '@disruptive-labs/client/react';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useState } from 'react';
import { AppClient, UserType } from '@app/api';

export const useAppAuth = () => useAuth<AppClient>();

export function useAuthRedirect(autoRedirectOnFirstRender = true) {
  const auth = useAppAuth();
  const router = useRouter();
  const [shouldRedirect, setShouldRedirect] = useState(false);

  const redirect = useCallback(() => {
    setShouldRedirect(true);
  }, []);

  // Handle auto-redirect on ready, if enabled
  useEffect(() => {
    if (!router.isReady || auth.isLoading) return;

    if (autoRedirectOnFirstRender && auth.isAuthenticated) redirect();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.isLoading, router.isReady, autoRedirectOnFirstRender, redirect]);

  // Handle manual redirect once auth.user available
  useEffect(() => {
    if (shouldRedirect) {
      setShouldRedirect(false);

      const userType = Object.values(UserType).find((item) => auth.user?.roles?.includes(item));
      router.replace(
        userType ? `/${userType !== UserType.Rsp ? userType : `ssp/case-managers`}/` : '/',
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.user, shouldRedirect]);

  return redirect;
}
