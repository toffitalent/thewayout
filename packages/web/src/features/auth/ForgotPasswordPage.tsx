import { Button, Field, Heading, Text, useForm } from '@disruptive-labs/ui';
import { useState } from 'react';
import { SEO } from '@app/components/SEO';
import { useAppDispatch, useAuthRedirect, useErrorIndicator } from '@app/hooks';
import { NextPageWithLayout } from '@app/types';
import { forgotPassword } from './actions';
import { AuthLayout } from './AuthLayout';

interface ForgotPasswordFormValues {
  username: string;
}

export const ForgotPasswordPage: NextPageWithLayout = function ForgotPasswordPage() {
  const dispatch = useAppDispatch();
  const showError = useErrorIndicator();
  const [successUsername, setSuccessUsername] = useState('');
  useAuthRedirect();

  const {
    control,
    submitForm,
    formState: { isSubmitting, isDirty },
  } = useForm({
    defaultValues: { username: '' },
    onSubmit: (values: ForgotPasswordFormValues) =>
      dispatch(forgotPassword(values.username))
        .unwrap()
        .then(() => {
          setSuccessUsername(values.username);
        })
        .catch((e) => {
          showError(e);
        }),
  });

  return (
    <>
      <SEO title="Forgot Password" description="" noIndex />
      {successUsername ? (
        <>
          <Heading as="h1" f="h2" mb={2.5}>
            Check your email
          </Heading>
          <Text mb={10}>
            We have sent an email to {successUsername}. Click the link in the email to reset your
            password. If you don&apos;t see the email, check your junk, social or other folders.
          </Text>
          <Button onClick={() => setSuccessUsername('')} variant="text">
            I didn&apos;t receive an email.
          </Button>
        </>
      ) : (
        <>
          <Heading as="h1" f="h2" mb={2.5}>
            Forgot Password
          </Heading>
          <Text mb={10}>Enter the email address associated with the account.</Text>
          <form onSubmit={submitForm}>
            <Field
              type="email"
              name="username"
              label="Email"
              placeholder="Your email address"
              validate={{ email: true }}
              fluid
              control={control}
            />
            <Button
              mt={10}
              type="submit"
              colorScheme="primary"
              disabled={isSubmitting || !isDirty}
              fluid
            >
              Send password reset
            </Button>
          </form>
        </>
      )}
    </>
  );
};

ForgotPasswordPage.getLayout = (page) => <AuthLayout>{page}</AuthLayout>;
