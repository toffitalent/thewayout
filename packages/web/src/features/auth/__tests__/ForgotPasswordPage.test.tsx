import userEvent from '@testing-library/user-event';
import { useRouter } from 'next/router';
import React from 'react';
import { useErrorIndicator } from '@app/hooks';
import { getByText, mockDispatch, mockThunks, render, waitFor } from '@test';
import * as actions from '../actions';
import { ForgotPasswordPage } from '../ForgotPasswordPage';

jest.mock('@app/hooks', () => ({
  ...jest.requireActual<any>('@app/hooks'),
  useErrorIndicator: jest.fn(),
}));

describe('ForgotPasswordPage', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockImplementation(() => ({
      isReady: true,
      query: {},
    }));
  });

  test('renders correctly', async () => {
    const { container } = render(<ForgotPasswordPage />);
    expect(container).toMatchSnapshot();
  });

  test('handles submission', async () => {
    const [forgotPassword] = mockThunks(actions, ['forgotPassword']);
    const { container, dispatch, getByRole, getByLabelText } = render(<ForgotPasswordPage />);

    await userEvent.type(getByLabelText('Email'), 'email@email.com', { delay: 1 });
    await userEvent.click(getByRole('button'));

    expect(forgotPassword).toBeCalledWith('email@email.com');
    expect(dispatch).toBeCalled();
    expect(container).toMatchSnapshot();
  });

  test('clears success message if email not received clicked', async () => {
    const { getByRole, getByLabelText, queryByText } = render(<ForgotPasswordPage />);

    await userEvent.type(getByLabelText('Email'), 'email@email.com', { delay: 1 });
    await userEvent.click(getByRole('button'));

    await waitFor(() => {
      expect(queryByText("I didn't receive an email.")).toBeInTheDocument();
    });

    await userEvent.click(getByRole('button'));

    await waitFor(() => {
      expect(
        queryByText('Enter the email address associated with the account.'),
      ).toBeInTheDocument();
    });
  });

  test('displays error on failure', async () => {
    const mockShowError = jest.fn();
    (useErrorIndicator as jest.Mock).mockImplementation(() => mockShowError);

    const error = new Error('Failed');
    const dispatch = mockDispatch(jest.fn().mockRejectedValue(error));
    const [forgotPassword] = mockThunks(actions, ['forgotPassword']);
    const { getByRole, getByLabelText } = render(<ForgotPasswordPage />, { dispatch });

    await userEvent.type(getByLabelText('Email'), 'email@email.com', { delay: 1 });
    await userEvent.click(getByRole('button'));

    expect(forgotPassword).toBeCalledWith('email@email.com');
    expect(dispatch).toBeCalled();
    expect(mockShowError).toBeCalledWith(error);
  });
});
