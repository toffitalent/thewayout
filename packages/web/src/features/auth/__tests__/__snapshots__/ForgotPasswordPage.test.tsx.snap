// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPasswordPage handles submission 1`] = `
<div>
  <h1
    class="text marginBottom-2-5 font-h2"
  >
    Check your email
  </h1>
  <p
    class="text marginBottom-10"
  >
    We have sent an email to 
    email@email.com
    . Click the link in the email to reset your password. If you don't see the email, check your junk, social or other folders.
  </p>
  <button
    class="button md text c-primary"
    type="button"
  >
    I didn't receive an email.
  </button>
</div>
`;

exports[`ForgotPasswordPage renders correctly 1`] = `
<div>
  <h1
    class="text marginBottom-2-5 font-h2"
  >
    Forgot Password
  </h1>
  <p
    class="text marginBottom-10"
  >
    Enter the email address associated with the account.
  </p>
  <form>
    <div
      class="control"
      role="group"
    >
      <label
        class="label"
        for=":r0:"
      >
        Email
      </label>
      <div
        class="root primary md fluid md"
      >
        <input
          class="input"
          id=":r0:"
          name="username"
          placeholder="Your email address"
          type="email"
          value=""
        />
      </div>
    </div>
    <button
      class="button md solid c-primary disabled fluid marginTop-10"
      disabled=""
      type="submit"
    >
      Send password reset
    </button>
  </form>
</div>
`;
